### PR TITLE
EZP-29253: Make session refresh re-try on timeouts and server errors

### PR DESCRIPTION
**Pull request by andrerom**
In order to try to improve stability of rest client, re-try session refresh which is often the one holding others back on timeout and 5xx errors. Do this for up to 180 seconds which can happen on slow deployment / security updates of backend, on cases where service needs to be taken down and cache needs to be re-generated.

NOTE: This does not solve issues when backend is down and client sends other requests to it, for instance when session has just been refreshed. However 1. depending on the call that will have to be identified on a case by case basis, can't handle it generically afaik. 2. v2 solves this.

Todo:
- [x] Rebase to right branch
- [x] In case of 5xx errors, maybe add some sleep / wait time in order to make sure we can survive a up to ~90~ 180 sec downtime / update / deployment window.

### CODE REVIEW

**Comment by natanael89** [comment]
I tested the solution. It correctly retries the session refresh requests. However, after three requests (if all of them fail), the end result is the same - the Editor will see the login form. The reporter mentioned (and I also think that it would be a good idea) that Users should not be presented with the login form while they are logged in and still have a valid session - even if the session refresh requests are failing. Perhaps a better idea would be not to retry the requests, but let them fail silently (ofc without refreshing the session)?

**Comment by andrerom** [comment]
> while they are logged in and still have a valid session 

Do we have anyway of knowing that? Or you suggest we just assume it.

>  let them fail silently (ofc without refreshing the session)?

We could do that after 3 attempts, only thing I see as a problem is that if this endpoint is down, all other endpoints are probably down, and typically calls to other endpoints are just waiting for session refresh to complete before they are sent by client to server.


Unless of course we manage to get more details from customer on cases where this fails which is not related to deployment / server being restarted and such which is what I'm trying to cover with timeout and 5xx error handling. If we get something we can reproduce on that, I would be much more comfortable changing the logic here.

**Comment by natanael89** [comment]
I was suggesting just assuming that the Editor still has a valid session. This is checked anyway on every request if I understand this correctly, so that shouldn't be an issue.
You are of course right in saying that if this endpoint is down, the other endpoints are also probably down. The reporter just seems to have an issue with showing the login form to the Editor when the errors happen (since the Editor is not really logged out).
In any case, I'll ask the reporter if retrying the session refresh requests is an acceptable solution for them and if not, I'll ask them to provide use case(s) not related to deployment.

**Comment by andrerom** [comment]
@adamwojs Updated

File: src/services/UserService.js [hunk 1]
```diff
@@ -1319,24 +1319,56 @@ define(['structures/SessionCreateStruct', 'structures/UserCreateStruct', 'struct
...
+     *                   NOTE: gets called every time, except on the first try!
      */
-    UserService.prototype.refreshSession = function (sessionId, callback) {
-        var that = this;
+    UserService.prototype.refreshSession = function (sessionId, callback, timeout, reTryCallback) {
+        var that = this, firstTry = true, date = new Date(), refreshSessionInfo, sessionRefresh = function(e, result) {
+            // We re-try session refresh for 180 seconds on timeouts/server errors as it is often the requests holding
+            // others back(see ConnectionManager), and is thus the one request we typically need to re-try a few
+            // times if backend has temporary downtime or is undergoing brief maintenance.
+            if (date - (new Date()) >= (timeout || 180000)) {
```
Comment (natanael89) [inline src/services/UserService.js]: Should be `(new Date()) - date` (other way around).

Comment (andrerom) [reply src/services/UserService.js]: Indeed 😅

**Comment by andrerom** [comment]
Updated again, cleaned up for review.

**Comment by andrerom** [comment]
Merging to make aware on beta tag.

**Pull request merged**